src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,252 @@ object RuleUtils {
+          .toSet -- index.allSourceFileInfos).toSeq.map(f => new Path(f.name))
+
+        if (filesAppended.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
+      // Remove sort order because we cannot guarantee the ordering of source files
+      val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+      object ExtractTopLevelPlanForShuffle {
+        type returnType = (LogicalPlan, Seq[Option[Attribute]], Boolean)
+        def unapply(plan: LogicalPlan): Option[returnType] = plan match {
+          case project @ Project(
+                _,
+                Filter(_, LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _))) =>
+            val indexedAttrs = getIndexedAttrs(project, index.indexedColumns)
+            Some(project, indexedAttrs, true)
+          case project @ Project(
+                _,
+                LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _)) =>
+            val indexedAttrs = getIndexedAttrs(project, index.indexedColumns)
+            Some(project, indexedAttrs, true)
+          case filter @ Filter(_, LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _)) =>
+            val indexedAttrs = getIndexedAttrs(filter, index.indexedColumns)
+            Some(filter, indexedAttrs, false)
+          case relation @ LogicalRelation(HadoopFsRelation(_, _, _, _, _, _), _, _, _) =>
+            val indexedAttrs = getIndexedAttrs(relation, index.indexedColumns)
+            Some(relation, indexedAttrs, false)
+        }
+        def getIndexedAttrs(
+            plan: LogicalPlan,
+            indexedColumns: Seq[String]): Seq[Option[Attribute]] = {
+          val attrMap = plan.output.attrs.map(attr => (attr.name, attr)).toMap
+          indexedColumns.map(colName => attrMap.get(colName))
+        }
+      }
+
+      // 2) Perform on-the-fly Shuffle with the same partition structure of index
+      // so that we could avoid incurring Shuffle of whole index data at merge stage.
+      // In order to utilize push-down filters, we would locate Shuffle node after
+      // Project or Filter node. (Case 1)
+      // However, if Project node excludes any of indexedColumns, Shuffle will be
+      // converted to RoundRobinPartitioning which can cause wrong result issues.
+      // So Shuffle should be located before Project node in that case. (Case 2)
+      // Case 1) Shuffle => Project => Filter => Relation (Project&Filter will be pushed down)
+      // Case 2) Project => Shuffle => Filter => Relation (Filter will be pushed down)
```

Comment:
Can you point me to the tests that are handling these two cases?

Reply:
I realized that we don't need to check this condition if we use Union for filter index rule here. 
For join rule, the index cannot be a candidate if the plan doesn't have all indexed columns.

Can I let this code without test and removed with #168 or add Union in this PR? Actually it's quite small change.